### PR TITLE
Add evolution bidder

### PR DESCRIPTION
**Pull request by And1sS**

### CODE REVIEW

File: src/test/resources/org/prebid/server/it/openrtb2/evolution/test-auction-evolution-request.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  "id": "tid",
+  "imp": [
+    {
+      "id": "impId001",
+      "banner": {
+        "w": 300,
+        "h": 250
+      },
+      "ext": {
+        "evolution": {
```
Comment (snahornyi) [inline src/test/resources/org/prebid/server/it/openrtb2/evolution/test-auction-evolution-request.json:11]: From `go` name should be `e_volution`

File: src/test/java/org/prebid/server/it/EvolutionTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                .header("X-Forwarded-For", "193.168.244.1")
+                .header("User-Agent", "userAgent")
+                .header("Origin", "http://www.example.com")
+                .body(jsonFrom("openrtb2/evolution/test-auction-evolution-request.json"))
+                .post("/openrtb2/auction");
+
+        // then
+        final String expectedAuctionResponse = openrtbAuctionResponseFrom(
+                "openrtb2/evolution/test-auction-evolution-response.json",
+                response, singletonList("evolution"));
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/it/EvolutionTest.java:45]: e_volution

File: src/main/resources/bidder-config/evolution.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      site-media-types:
+        - banner
+        - video
+        - native
+      supported-vendors:
+      vendor-id: 957
+    usersync:
+      url: https://sync.e-volution.ai/pbserver?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&ccpa={{us_privacy}}&redirect=
+      redirect-url: /setuid?bidder=e_volution&gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&uid=[UID]
+      cookie-family-name: evolution
```
Comment (snahornyi) [inline src/main/resources/bidder-config/evolution.yaml:25]: e_volution

File: src/main/resources/bidder-config/evolution.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+adapters:
+  evolution:
```
Comment (snahornyi) [inline src/main/resources/bidder-config/evolution.yaml:2]: e_volution

File: src/main/java/org/prebid/server/spring/config/bidder/EvolutionConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.PropertySource;
+
+import javax.validation.constraints.NotBlank;
+
+@Configuration
+@PropertySource(value = "classpath:/bidder-config/evolution.yaml", factory = YamlPropertySourceFactory.class)
+public class EvolutionConfiguration {
+
+    private static final String BIDDER_NAME = "evolution";
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/spring/config/bidder/EvolutionConfiguration.java:24]: e_volution

File: src/main/java/org/prebid/server/bidder/evolution/EvolutionBidder.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    private BidType getBidMediaType(Bid bid) {
+        try {
+            final EvolutionBidResponse evolutionBidResponse = mapper.mapper().convertValue(
+                    bid.getExt(), EvolutionBidResponse.class);
+            return evolutionBidResponse != null ? evolutionBidResponse.getMediaType() : BidType.banner;
+        } catch (IllegalArgumentException e) {
+            return BidType.banner;
+        }
+    }
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/evolution/EvolutionBidder.java:95]: Let's prepare some updates for this methods
```java
 private List<BidderBid> bidsFromResponse(BidResponse bidResponse) {
        final SeatBid firstSeatBid = bidResponse.getSeatbid().get(0);
        return CollectionUtils.emptyIfNull(firstSeatBid.getBid()).stream()
                .filter(Objects::nonNull)
                .map(bid -> BidderBid.of(bid, getBidMediaType(bid.getExt()), bidResponse.getCur()))
                .collect(Collectors.toList());
    }

    private BidType getBidMediaType(JsonNode bidExt) {
        final JsonNode mediaTypeNode = bidExt != null ? bidExt.get("mediaType") : null;
        try {
            return ObjectUtils.defaultIfNull(mapper.mapper()
                    .convertValue(mediaTypeNode, BidType.class), BidType.banner);
        } catch (IllegalArgumentException e) {
            return BidType.banner;
        }
    }
```
Then EvolutionBidResponse class can be deleted

File: src/main/java/org/prebid/server/bidder/evolution/EvolutionBidder.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+    private HttpRequest<BidRequest> createRequest(BidRequest request) {
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(HttpUtil.headers())
+                .payload(request)
+                .body(mapper.encode(request))
+                .build();
+    }
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/evolution/EvolutionBidder.java:55]: Let's get rid of this method and put it's logic directly to `makeHttpRequests` method

File: src/main/java/org/prebid/server/bidder/evolution/EvolutionBidder.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+
+    public EvolutionBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/evolution/EvolutionBidder.java:43]: Also pls add test that only one request should be created for all imps

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/evolution/model/EvolutionBidResponse.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package org.prebid.server.bidder.evolution.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+
+@Value
+@AllArgsConstructor(staticName = "of")
+public class EvolutionBidResponse {
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/bidder/evolution/model/EvolutionBidResponse.java:10]: Do we need this model ?

**Review by rpanchyk — CHANGES_REQUESTED** [review]

File: src/main/resources/bidder-config/evolution.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        - native
+      site-media-types:
+        - banner
+        - video
+        - native
+      supported-vendors:
+      vendor-id: 957
+    usersync:
+      url: https://sync.e-volution.ai/pbserver?gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&ccpa={{us_privacy}}&redirect=
+      redirect-url: /setuid?bidder=e_volution&gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&uid=[UID]
```
Comment (snahornyi) [inline src/main/resources/bidder-config/evolution.yaml:24]: In redirect-url all this parameters should be present
`&gdpr={{gdpr}}&gdpr_consent={{gdpr_consent}}&us_privacy={{us_privacy}}`

**Review by rpanchyk — APPROVED** [review]

**Review by snahornyi — APPROVED** [review]

**Pull request merged**